### ISSUE TITLE
new NSM forwarding logic breaks Mockito

### ISSUE THREAD

**Issue by srawlins** [dart-lang/sdk#33031]
Open items to move forward:

- [x] Landing Mockito fix [@srawlins]
- [x] Rolling Mockito in Flutter [@srawlins] (https://github.com/flutter/flutter/pull/17442)
- [ ] Fixing ~10 internal tests [@srawlins]
- [x] Landing CFE implementation [@sjindel-google]
- [x] Landing DDC implementation. [@jmesserly]
- [ ] Communicating / updating the spec [@lrhn @leafpetersen]

@sjindel-google 's 7d5025e814fa4716fb8c6a4c104ea1c28d578911 breaks how Mockito works. Given

```dart
class Cat {
  bool eatFood(String food, {bool hungry}) => true;
}

// Mock class
class MockCat extends Mock implements Cat {}

// mock creation
var cat = new MockCat();
```

The user should now be able to stub out `eatFood` like so:

```dart
when(cat.eatFood("bird")).thenReturn(true);
```

However, Mock's noSuchMethod receives something different after 7d5025e814fa4716fb8c6a4c104ea1c28d578911: it used to receive an Invocation with 1 positional argument, and 0 named arguments. Now it receives the 1 positional argument, and the 1 named argument (with the default value specified by the parameter). The _presence_ of arguments or lack thereof, in the Invocation passed to noSuchMethod is core to Mockito's implementation.

This is blocking the current Flutter roll (https://github.com/flutter/flutter/pull/17230), as Flutter has some affected tests. This will also block the SDK roll into internal Google, as we use Mockito.

I also imagine this is just... a big breaking change to how NSM works. I wouldn't be surprised if this broke other tools that depend on NSM functionality.

I think this change must be rolled back ASAP, to unblock Flutter. We can then discuss whether the change is possible for Dart 2.

Related to #33019

CC @sjindel-google @keertip @tvolkert @yjbanov

**Comment by keertip** [comment]
cc @kmillikin @mraleph

**Comment by aam** [comment]
Samir actually reverted 7d5025e on https://github.com/dart-lang/sdk/commit/24dd9b4176ec9d8694ba89cfa07e503f9c53a35d, but that broke the buildbots blocking further commits and there was no clear reason for the revert,  so to unblock the buildbots I re-reverted it on https://github.com/dart-lang/sdk/commit/9a7e1f64a2e1f51464dbb5c1ee1dab2f375aa23c.

**Comment by aam** [comment]
Closed via 2765fcf

**Comment by srawlins** [comment]
Thanks, @aam

**Comment by lrhn** [comment]
The behavior of the `noSuchMethod` forwarder forwarding all declared parameters is correct and what we want.

I'm not sure what causes Mockito's issue here - it should get equivalent `Invocation`s both when configuring and when using the mock, so it should still be able to match them up and return the expected value.

It is correct that you can no longer make a mock act differently depending on whether an optional parameter is omitted, or it is provided with the same value as the default value. A real Dart method could never detect that (apart from a short period where we didn't know better), so the mock should still be perfectly capable of emulating any valid behavior.

So, we will have to reapply this fix to make nSM forwarders do the correct thing.

**Comment by kmillikin** [comment]
This is complicated to follow.  I think this is what happened in order:

1. https://github.com/dart-lang/sdk/commit/7d5025e814fa4716fb8c6a4c104ea1c28d578911 landed giving us the intended behavior but breaking Mockito
1. https://github.com/dart-lang/sdk/commit/24dd9b4176ec9d8694ba89cfa07e503f9c53a35d reverted that for `<INSERT REASONING HERE>`, giving us the incorrect behavior that Mockito relies on
1. https://github.com/dart-lang/sdk/commit/9a7e1f64a2e1f51464dbb5c1ee1dab2f375aa23c reverted that because it broke the presubmit buildbots, giving us the intended behavior again and breaking Mockito
1. https://github.com/dart-lang/sdk/commit/2765fcf2aecd3841d082fedaeafc00a73a965f8c reverted that because it broke Mockito, giving us the incorrect behavior again

We do intend to implement the intended behavior (almost by definition) so I think we will have to change Mockito.

**Comment by srawlins** [comment]
I'll look into how Mockito can handle this fix. I expect that the code of Mockito _users_ will have to change, which means we'll need a few weeks [or months](https://en.wikipedia.org/wiki/Hofstadter%27s_law) runway. I'll comment here when I know more.

**Comment by dgrove** [comment]
Also note that such a breaking change would block rolls of the SDK into google3 (and hence Flutter) until those changes in user code could be made (ie, I'm not sure at all that this change can be made for Dart 2).

**Comment by srawlins** [comment]
Honest question, was 7d5025e written because like, some package that uses `noSuchMethod` benefits from the change, or did we just notice that nSM was being handled wrong, and we don't want to be wrong?

I think we have a workable solution for Mockito, but it definitely makes Mockito less usable. Is there a net benefit? Does some other package become more usable, or possible?

**Comment by tvolkert** [comment]
My 2c: this change seems an obviously correct one that makes things noticeably more correct.  The only question is how to feasibly roll it out.

**Comment by mraleph** [comment]
This should be classified as a language change and the decision is then a language team decision. 

If migration is unlikely - then we should amend Dart 2 specification making the current behavior "the specified" behavior.

/cc @leafpetersen @lrhn

**Comment by jmesserly** [comment]
FYI -- I've had to be careful in DDC to keep mockito working for SDK rolls. DDC always preserves the provided arguments for that reason, rather than filling in unnecessary default values (contrary to what #33019 asks DDC to do).

**Comment by yjbanov** [comment]
@lrhn, @leafpetersen I'm curious, why is there desire to have `noSuchMethod` behave like a "real Dart method"? `noSuchMethod` already has some crazy extra-linguistic capabilities that almost feel like runtime reflection or monkey patching. I think most of its usefulness comes precisely from being able to do crazy things that "real Dart" can't do. To me `noSuchMethod` is a way to _reflect_ on a function call, which, among other things should be able to tell me which arguments were passed.

**Comment by natebosch** [comment]
Can someone from the language team please clarify what the benefits of these semantics are? They are not obviously correct to me.

Why is a default argument value a property of a method invocation? It seems to me a property of a method implementation.

If a class extends another it doesn't see the super-class default values for arguments. Why is `noSuchMethod` different from other ways a method can be overridden?

**Comment by leafpetersen** [comment]
> Can someone from the language team please clarify what the benefits of these semantics are? They are not obviously correct to me.
>
> Why is a default argument value a property of a method invocation? It seems to me a property of a method implementation.

I don't think this really has anything to do with whether or not default arguments are properties of method invocation signatures.  The primary concern (in so far as it is a concern) is that it is otherwise un-observable whether or not an optional parameter is passed or not passed, and this is a bit of a bad smell.  Example:

```
class Foo {
  void m([int x]) {
     // Write the code here to detect whether `x` was passed explicitly or not.
  }
}
```

As intended with this feature, there is no implementation of `m`.  With the behavior that mockito is relying on, there is such an implementation: namely, leave `m` abstract and implement it in NSM.  It's an awful way to get that behavior though.  If we want to provide that power, it seems better to just provide it directly, since we're requiring implementations to support it anyway.

It isn't the end of the world if we stay with the current semantics, but I think it is reasonable to ask for a thorough consideration of alternatives.  I believe that @srawlins is exploring options in mockito.

**Comment by natebosch** [comment]
I'm not sure I follow. Are we saying that we believe it's impossible to know whether a parameter was passed explicitly and we want it to remain impossible even with NSM because we're afraid too many people will use NSM to gain this capability and we don't want the capability to exist at all?

If we allow the implementation in extending classes there is a workaround which I've seen in the wild.

```dart
const _placeholder = const Object();

abstract class Foo {
  void m([int x]);
}

class Bar extends Foo {
  @override
  void m([Object x = _placeholder]) {
    // `x` was passed explicitly if it is != _placeholder
  }
}
```

**Comment by leafpetersen** [comment]
@natebosch Fair point that you can do this with markers if you weaken the signature.  You are, of course, implementing a different (weaker) signature.

In any case, I don't think anyone feels that this is a feature that makes or breaks Dart 2.  Our intention with the NSM forwarders was that they simply fill in boilerplate that you could have written yourself, and optional argument presence forwarding breaks that abstraction.  Not the end of the world either way.

**Comment by vsmenon** [comment]
I don't see the cycles to do this migration right now.  Unless @srawlins has discovered otherwise, this looks to be substantial work.

I suggest spec'ing the working behavior for Dart 2 and revisiting this (if desired) afterward in conjunction with needed Mockito redesign.

@leafpetersen @lrhn - are you okay with that?

**Comment by leafpetersen** [comment]
If we don't have the cycles we don't have the cycles.

**Comment by vsmenon** [comment]
Quick update: @srawlins and @jmesserly are trying a presubmit with the new semantics.  This should tell us the migration cost.  It appears that implementation cost is low (as most of it is already done).

**Comment by dgrove** [comment]
@srawlins and @jmesserly can you provide an update today on this?

**Comment by srawlins** [comment]
I'm going to presubmit the new semantics in DDC tonight. Basically what @vsmemnon [said](https://github.com/dart-lang/sdk/issues/33031#issuecomment-387520908).

**Comment by srawlins** [comment]
OK, update for the night: things mostly look good. 😅 

Jenny wrote a prototype that seemed very very functional for DDC. That was necessary to get a true sense of the scale of any cleanup that might be needed internally. It looks like very few tests actually start failing. Of the ones that do, it looks like they might be failing just because of a small bug in Jenny's implementation (given a function like `fn(int a, {SomeClass b = const SomeClass("SOME_STRING")})`, Mockito claimed that a call like `fn(7)` was not equivalent to a call like `fn(7)` because `b` wasn't equals to `b` 😕 ). Given all that, I think we're a go on this change (just my 2c), after I land a fix for https://github.com/dart-lang/mockito/issues/122, which looks like is backwards compatible.

I'd like to test Jenny's prototype with the const object bit fixed though, just to really know exactly how big the cleanup is.

**Comment by dgrove** [comment]
Good news! sounds like @sjindel-google should reland this change, right @srawlins ?

**Comment by srawlins** [comment]
I'd prefer to land a fix for https://github.com/dart-lang/mockito/issues/122 first, to prevent another flutter breakage during their role. I have a fix mostly ready. I can mail and land it tomorrow.

**Comment by leafpetersen** [comment]
I made some small tweaks to @jmesserly's cl and got 2 previously failing tests running, I'll try to run a global overnight with my tweaks.

**Comment by jmesserly** [comment]
I also figured out what was wrong with my CL (forgot to pass `isConst: true` after much refactoring to DDC) and am testing the cleaned up version.

**Comment by tvolkert** [comment]
Just to make sure I understand the current state of things: it sounds like we're going to be able to re-land 7d5025e _unmodified_ because we found viable workarounds in mockito (that comes with a workable migration path) and DDC?

**Comment by jmesserly** [comment]
> Just to make sure I understand the current state of things: it sounds like we're going to be able to re-land 7d5025e unmodified because we found viable workarounds in mockito (that comes with a workable migration path) and DDC?

I think it will be reverting https://dart-review.googlesource.com/c/sdk/+/53540 ... it looks like there maybe have been some small tweaks to the CL to keep buildbots happy, but otherwise I presume it's almost the same code as https://github.com/dart-lang/sdk/commit/7d5025e814fa4716fb8c6a4c104ea1c28d578911, yes. mockito has a workaround and DDC has an implementation (we're still testing, but it seems like it's close)

**Comment by sjindel-google** [comment]
Unfortunately we can't land 7d5025e just yet due to a few failures on precompiled bots. Please let me know when I can re-land it with fixes.

**Comment by sjindel-google** [comment]
Yes, the fix is ready and the revision is approved. I'm just wait for the go-ahead.

**Comment by vsmenon** [comment]
The internally tested CL (195908155) includes @srawlins's Mockito fix and @jmesserly's DDC implementation.  So, rolling forwarding means:

[ ] Landing @sjindel-google 's CFE implementation.
[ ] Landing @jmesserly 's DDC implementation.
[ ] Landing @srawlins 's Mockito fix.
[ ] Fixing ~10 internal tests (@srawlins )
[ ] Communicating / updating the spec (@lrhn @leafpetersen )

Are we missing anything?  Is there a chance something will break in Flutter or internal Flutter apps?  Are the VM AOT/JIT and Dart2JS all covered by Samir's fix?

**Comment by vsmenon** [comment]
Yes, I think we'll need the Mockito fix landed and rolled into Flutter and internal.

**Comment by srawlins** [comment]
Agree with everything above. @tvolkert or @yjbanov might correct the Flutter bit; it could be that all of Flutter was cleaned up in that fun roll last week. Otherwise, we should land and release mockito fix, and bump mockito in flutter. I can do these things today.

**Comment by vsmenon** [comment]
Thanks!  I added check boxes to the initial comment at top so we can track.

**Comment by vsmenon** [comment]
@srawlins - are we still waiting on the flutter roll?

**Comment by srawlins** [comment]
Yeah https://github.com/flutter/flutter/pull/17487

**Comment by srawlins** [comment]
Landed in flutter. nSM Forwarding is cool to land in the SDK.

**Comment by vsmenon** [comment]
Thanks, Sam!

@sjindel-google and @jmesserly - can you please land your impl CLs?

**Comment by jmesserly** [comment]
> @jmesserly - can you please land your impl CLs?

yeah I tried it failed for a status issue (it needs test fixes from @sjindel-google 's CL). Trying CQ again w/ that test temporarily marked failing in DDC (we can mark passing once CFE CL lands).

**Comment by keertip** [comment]
@jmesserly , see quite a few test failures in DDC internally after your [commit](https://github.com/dart-lang/sdk/commit/7359d4ee10302827ebcab1ca390563b110d74427), testing at commit `4b239de`, previous tested at `4a323f8`.

`
Invalid argument(s): A typed argument was passed in as a named argument named "Symbol("cancelOnError")", but did not pass a value for `named`. Each typed argument that is passed as a named argument needs to specify the `named` argument. For example: `when(obj.fn(x: typed(any, named: "x")))`.
`

@srawlins, do we need an update to mockito package to fix these?

**Comment by srawlins** [comment]
@keertip Yes we need to update mockito to fix those.

**Comment by vsmenon** [comment]
@sjindel-google - yes, please.  :-)

**Comment by vsmenon** [comment]
@sjindel-google and @jmesserly - are we done on the implementation side?

**Comment by sjindel-google** [comment]
Yes.

**Comment by vsmenon** [comment]
@srawlins - are you unblocked from fixing those ~10 tests?

@leafpetersen @lrhn - how should this change be communicated?

**Comment by srawlins** [comment]
Yes, I can fix them when we roll >=dev.56 internally.

**Comment by vsmenon** [comment]
@srawlins - thanks!

@leafpetersen is working on how to communicate this as part of overall NSM changes.  I'll go ahead and close this bug.
